Week-3/MemoizationAssignment/memoization.js

```diff
@@ -0,0 +1,38 @@
+/** Normal Sum Function 
+ * 
+ * It will perform calculation every single time
+*/
+
+function add(...args) {
+    let sum = 0;
+    args.forEach((num) => sum += num);
+    return sum;
+}
+
+/** Memoized Sum Function */
+
+function memoize(add) {
+    const cache = new Map();
+    return function (...args) {
+        const key = args.join("-");
+        // if already present return that
+        if (cache.has(key)) {
+            console.log("used cache");
+            return cache.get(key)
+        }
+        // if not present, call the function to calculate sum
+        // and then save sum for the new key in cache
+        console.log("calculated this");
+        cache.set(key, add(...args));
+        return cache.get(key);
```

Comment:
doesn't it increases complexity, as cache.get(key) will search entire cache for result, when the result is already called by the function.